compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt

```diff
@@ -75,39 +73,69 @@ object FirDestructuringDeclarationChecker : FirPropertyChecker() {
                 else -> null
             } ?: return
 
-        when (val diagnostic = reference.diagnostic) {
-            is ConeUnresolvedNameError -> {
-                reporter.report(
-                    FirErrors.COMPONENT_FUNCTION_MISSING.on(
-                        originalDestructuringDeclarationOrInitializerSource,
-                        diagnostic.name,
-                        originalDestructuringDeclarationType
-                    ),
-                    context
-                )
-            }
-            is ConeAmbiguityError -> {
-                reporter.report(
-                    FirErrors.COMPONENT_FUNCTION_AMBIGUITY.on(
-                        originalDestructuringDeclarationOrInitializerSource,
-                        diagnostic.name,
-                        diagnostic.candidates
-                    ),
-                    context
-                )
-            }
-            is ConeInapplicableCandidateError -> {
-                if (originalDestructuringDeclarationType.isNullable) {
-                    reporter.report(
-                        FirErrors.COMPONENT_FUNCTION_ON_NULLABLE.on(
-                            originalDestructuringDeclarationOrInitializerSource,
-                            (diagnostic.candidate.symbol.fir as FirSimpleFunction).name
-                        ),
+        when (val reference = componentCall.calleeReference) {
+            is FirResolvedNamedReference -> {
+                val destructuringCall = (reference.resolvedSymbol.fir as? FirSimpleFunction) ?: return
+                val destructuringType = destructuringCall.returnTypeRef.coneType
+                if (destructuringType is ConeKotlinErrorType) {
+                    // There will be other errors on this error type.
+                    return
+                }
+                if (destructuringType is ConeTypeParameterType) {
```

Comment:
This doesn't seem to work for the following code. `i:T` should be highlighted inside the call of `foo` in `bar`.

```
data class A<T>(val i: T)

fun <T> foo(block: (A<T>) -> Unit) {}

fun <T, R> bar() {
    foo<R> { (i: T) ->
    }
}
```

Reply:
Oops! Now this comment is redundant with https://github.com/JetBrains/kotlin/pull/4286#discussion_r608162295

Reply:
Yep, same concern. Added your example as well, thanks!